pkg/clients/keyvault/secret/keyvault_secret.go

```diff
@@ -0,0 +1,205 @@
+func GenerateAttributes(spec *v1alpha1.KeyVaultSecretAttributesParameters) *keyvault.SecretAttributes {
+	if spec == nil {
+		return nil
+	}
+
+	return &keyvault.SecretAttributes{
+		Enabled:   spec.Enabled,
+		NotBefore: metav1TimeToUnixTime(spec.NotBefore),
+		Expires:   metav1TimeToUnixTime(spec.Expires),
+	}
+}
+
+// ExtractSecretValue extracts secret value from secretRef.
+func ExtractSecretValue(ctx context.Context, client client.Client, value *xpv1.SecretKeySelector) (string, error) {
+	ref := xpv1.CommonCredentialSelectors{SecretRef: value}
+	val, err := resource.ExtractSecret(ctx, client, ref)
+
+	return string(val), err
+}
+
+// LateInitialize fills the spec values that user did not fill with their
+// corresponding value in the Azure, if there is any.
+func LateInitialize(spec *v1alpha1.KeyVaultSecretParameters, az keyvault.SecretBundle) {
+	spec.Tags = azure.LateInitializeStringMap(spec.Tags, az.Tags)
+	spec.ContentType = azure.LateInitializeStringPtrFromPtr(spec.ContentType, az.ContentType)
+	spec.SecretAttributes = lateInitializeSecretAttributes(spec.SecretAttributes, az.Attributes)
+}
+
+func lateInitializeSecretAttributes(sa *v1alpha1.KeyVaultSecretAttributesParameters, az *keyvault.SecretAttributes) *v1alpha1.KeyVaultSecretAttributesParameters {
+	if az == nil {
+		return sa
+	}
+	if sa == nil {
+		sa = &v1alpha1.KeyVaultSecretAttributesParameters{}
+	}
+	sa.Expires = lateInitializeTimePtrFromUnixTimePtr(sa.Expires, az.Expires)
+	sa.NotBefore = lateInitializeTimePtrFromUnixTimePtr(sa.NotBefore, az.NotBefore)
+	sa.Enabled = azure.LateInitializeBoolPtrFromPtr(sa.Enabled, az.Enabled)
+	return sa
+}
+
+func lateInitializeTimePtrFromUnixTimePtr(mt *metav1.Time, ut *date.UnixTime) *metav1.Time {
+	if mt != nil {
+		return mt
+	}
+	return unixTimeToMetav1Time(ut)
+}
+
+func unixTimeToMetav1Time(t *date.UnixTime) *metav1.Time {
+	if t == nil {
+		return nil
+	}
+	return &metav1.Time{Time: time.Time(*t)}
+}
+
+func metav1TimeToUnixTime(t *metav1.Time) *date.UnixTime {
+	if t == nil {
+		return nil
+	}
+
+	return (*date.UnixTime)(&t.Time)
+}
+
+func overrideParameters(sp v1alpha1.KeyVaultSecretParameters, sb keyvault.SecretBundle, val string) keyvault.SecretBundle {
+	sb.Value = azure.ToStringPtr(val)
+
+	if sp.ContentType != nil {
+		sb.ContentType = sp.ContentType
+	}
+
+	if sp.Tags != nil {
+		if sb.Tags == nil {
+			sb.Tags = azure.ToStringPtrMap(sp.Tags)
+		} else {
+			for k, v := range sp.Tags {
+				sb.Tags[k] = azure.ToStringPtr(v)
+			}
+		}
+	}
+
+	if sp.SecretAttributes != nil {
+		attr := GenerateAttributes(sp.SecretAttributes)
+		if sb.Attributes == nil {
+			sb.Attributes = attr
+		} else {
+			sb.Attributes.Enabled = attr.Enabled
+			sb.Attributes.NotBefore = attr.NotBefore
+			sb.Attributes.Expires = attr.Expires
+		}
+	}
+
+	return sb
+}
+
+func unixTimeCopier(v interface{}) (interface{}, error) {
+	// Just... copy it.
+	return v.(date.UnixTime), nil
```

Comment:
If this is an identity function for `date.UnixTime`, and because we register it for the type `date.UnixTime`, do we really need this? Please also see the discussion on the need to make a copy of the observed `keyvault.SecretBundle`.